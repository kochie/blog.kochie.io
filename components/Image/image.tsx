import React, { useEffect, useRef, CSSProperties } from 'react'

export interface Image {
    lqip: string,
    src: string,
    width: number | string,
    height: number | string,
    style?: CSSProperties
}

import imageStyle from "./image.less"

export default ({lqip, src, width, height, style}: Image) => {
    const imgRef = useRef<HTMLImageElement>(null)

    async function getImage() {
        try{
            const image = await fetch(src).catch(error => {throw error})

            if (!image.ok) return

            const blob = await image.blob()
            
            if (imgRef.current) {
                imgRef.current.src = URL.createObjectURL(blob)
                imgRef.current.className = imageStyle.sharpen
            }
        } catch (error) {
            console.error(error)
        }
    }

    useEffect(() => {
        setTimeout(() => getImage(), 1000)
    })

    return (
        <div className={imageStyle.container} style={{...style, width, height}}>
            <img ref={imgRef} src={lqip} width={"100%"} height={"100%"} className={imageStyle.blur} />
        </div>
    )
}