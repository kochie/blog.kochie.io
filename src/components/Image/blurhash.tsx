import React from 'react'

export const blurhash = (): React.ReactElement => {
  return <canvas />
}
