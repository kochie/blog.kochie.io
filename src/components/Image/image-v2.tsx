import React from 'react'

interface ImageProps {
  src: string
  width?: number
  height?: number
  hash?: string
}

const Image = ({ src, width, height }: ImageProps): React.ReactElement => {
  return (
    <>
      <img src={src} width={width} height={height} />
    </>
  )
}

export const makeHash = (src: string): string => {
  console.log(src)
  return src
}

export default Image
