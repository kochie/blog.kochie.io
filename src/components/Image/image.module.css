.blur {
  filter: blur(10px);
  object-fit: cover;
}

.container {
  overflow: hidden;
}

.sharpen {
  object-fit: cover;
  animation: sharpen 0.5s both;
}

.image {
  object-fit: cover;
  object-position: center;
  width: 100%;
  height: 100%;
}

@keyframes sharpen {
  from {
    filter: blur(10px);
  }
  to {
    filter: blur(0px);
  }
}
