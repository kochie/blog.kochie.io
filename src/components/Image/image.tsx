import React, { useEffect, useRef, CSSProperties } from 'react'
import imageStyle from '../../styles/image.module.scss'

export interface Image {
  lqip: string
  src: string
  width?: number | string
  height?: number | string
  alt?: string
  style?: CSSProperties
  className?: string
  loadOnObserve?: boolean
}

export default ({
  lqip,
  src,
  width,
  height,
  style,
  alt,
  className = '',
  loadOnObserve = false,
}: Image) => {
  const imgRef = useRef<HTMLImageElement>(null)

  async function getImage() {
    // console.log("WOAH")
    try {
      const image = await fetch(src).catch((error) => {
        throw error
      })

      if (!image.ok) return

      const blob = await image.blob()

      if (imgRef.current) {
        imgRef.current.src = URL.createObjectURL(blob)
        imgRef.current.className = imageStyle.sharpen
      } else {
        console.warn('imgRef does not have a current reference!')
      }
    } catch (error) {
      console.error(error)
    }
  }

  function createIntersectionObserver(): IntersectionObserver {
    const observer = new IntersectionObserver((entries, observer) => {
      // console.log(entries.length)
      entries.forEach((entry) => {
        if (entry.isIntersecting && imgRef.current) {
          getImage()
          observer.unobserve(imgRef.current)
        }
      })
    })
    if (imgRef.current) observer.observe(imgRef.current)
    return observer
  }

  useEffect(() => {
    if (!loadOnObserve) getImage()
    else {
      const observer = createIntersectionObserver()
      return () => {
        if (imgRef.current) observer.unobserve(imgRef.current)
      }
    }
  })

  const containerClasses = [imageStyle.container, className].join(' ')

  return (
    <div className={containerClasses} style={{ ...style, width, height }}>
      <img
        ref={imgRef}
        src={lqip}
        width={'100%'}
        height={'100%'}
        className={imageStyle.blur}
        alt={alt}
      />
    </div>
  )
}
