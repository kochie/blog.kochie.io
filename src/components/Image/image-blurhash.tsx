import { useEffect, useRef, CSSProperties } from 'react'
// import { Blurhash } from "./blurhash";
import { decode } from 'blurhash'
// import Jimp from 'jimp'
import imageStyle from '../../styles/image.module.scss'

export interface ImageProps {
  lqip: string
  src: string
  width?: number | string
  height?: number | string
  alt?: string
  style?: CSSProperties
  className?: string
  loadOnObserve?: boolean
}

const BlurhashImage = ({
  lqip,
  src,
  width,
  height,
  style,
  alt,
  className = '',
  loadOnObserve = false,
}: ImageProps): React.ReactElement => {
  const imgRef = useRef<HTMLImageElement>(null)
  const canvasRef = useRef<HTMLCanvasElement>(null)

  async function getImage(): Promise<void> {
    try {
      const image = await fetch(src)

      if (!image.ok) {
        console.warn(`image failed to load src: ${src}`)
        return
      }

      const blob = await image.blob()

      if (imgRef.current) {
        imgRef.current.src = URL.createObjectURL(blob)
        imgRef.current.className = imageStyle.sharpen
      } else {
        console.warn('imgRef does not have a current reference!')
      }
    } catch (error) {
      console.error(error)
    }
  }

  function createIntersectionObserver(): IntersectionObserver {
    const observer = new IntersectionObserver((entries, observer) => {
      // console.log(entries.length)
      entries.forEach((entry) => {
        if (entry.isIntersecting && imgRef.current) {
          getImage()
          observer.unobserve(imgRef.current)
        }
      })
    })
    if (imgRef.current) observer.observe(imgRef.current)
    return observer
  }

  useEffect(() => {
    if (!loadOnObserve) getImage()
    else {
      const observer = createIntersectionObserver()
      return (): void => {
        if (imgRef.current) observer.unobserve(imgRef.current)
      }
    }
  })

  // useEffect(() => {

  // }, [])

  const pixels = decode(lqip, 32, 32)
  const imageBlob = new Blob([pixels.buffer], { type: 'image/png' })
  const imageUrl = URL.createObjectURL(imageBlob)

  const containerClasses = [imageStyle.container, className].join(' ')

  return (
    <div className={containerClasses} style={{ ...style, width, height }}>
      <canvas ref={canvasRef} width={width} height={height} />
      <img
        ref={imgRef}
        src={imageUrl}
        width={'100%'}
        height={'100%'}
        className={imageStyle.blur}
        alt={alt}
      />
    </div>
  )
}

export default BlurhashImage
