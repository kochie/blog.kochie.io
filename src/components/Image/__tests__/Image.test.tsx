import React from 'react'
import { create } from 'react-test-renderer'

import Image from '../image-v2'

it('renders correctly', () => {
  const tree = create(<Image lqip={'lqip'} src={'source'} />).toJSON()
  expect(tree).toMatchSnapshot()
})
