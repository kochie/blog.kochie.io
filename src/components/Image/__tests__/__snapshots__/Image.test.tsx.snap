// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "overflow": "hidden",
      "width": "100%",
    }
  }
>
  <img
    className="image blur"
    src="lqip"
  />
</div>
`;
